Sources/PersistentCollections/PersistentDictionary._Node.swift

```diff
@@ -478,26 +481,35 @@ extension PersistentDictionary._Node: _DictionaryNodeProtocol {
   ) -> _Node {
     assert(isCollisionNode)
 
-    let content: [Element] = Array(self)
-    let hash = _HashValue(content.first!.key)
-
+    let hash = _HashValue(_items.first!.key)
     guard path._hash == hash else {
       effect.setModified()
       return _mergeKeyValPairAndCollisionNode(item, path, self, hash)
     }
 
-    if let offset = content.firstIndex(where: { item.key == $0.key }) {
-      var updatedContent: [Element] = []
-      updatedContent.reserveCapacity(content.count + 1)
-      updatedContent.append(contentsOf: content[0 ..< offset])
-      updatedContent.append(item)
-      updatedContent.append(contentsOf: content[(offset + 1)...])
-      effect.setReplacedValue(previousValue: content[offset].value)
-      return _Node(/* hash, */ collisions: updatedContent)
+    if let offset = _items.firstIndex(where: { $0.key == item.key }) {
+      effect.setReplacedValue(previousValue: _items[offset].value)
+      let node = isUnique ? self : self.copy()
+      node._mutableItems[offset] = item
+      return node
+    }
+
+    let hasRoomForItem = itemCount < itemCapacity
+    let dst: _Node
+    if isUnique && hasRoomForItem {
+      dst = self
     } else {
-      effect.setModified()
-      return _Node(/* hash, */ collisions: content + [item])
+      dst = self.copy(itemCapacityGrowthFactor: hasRoomForItem ? 1 : 2)
     }
+    assert(dst.itemCount < dst.itemCapacity)
+    let count = dst.itemCount
+    _rangeInsert(item, at: count, into: dst.itemBaseAddress, count: count)
+    dst.header.itemMap = _Bitmap(bitPattern: count + 1)
+    dst.header.childMap = dst.header.itemMap
```

Comment:
Should we assert the hash collision node header invariant?

Or introduce a domain specific setter that capture the intent that both bitmaps are identical and store the count?

Reply:
Oh good idea, will do.